docs/src/lib/types.md

```diff
@@ -72,6 +72,10 @@ that could perform a transformation without copying the columns). Examples of su
 [`hcat`](@ref), [`filter`](@ref), [`dropmissing`](@ref), `getindex`,
 [`copy`](@ref) or the [`DataFrame`](@ref) constructor mentioned above.
 
+The generic single-argument constructor `DataFrame(table)` has `copycols=nothing` by default, where
+certain Tables.jl-compatible inputs can signal that a copy of columns doesn't need to be made (this
+is done by wrapping the source table in `Tables.CopiedColumns`). [CSV.jl](https://csv.juliadata.org/stable) does this when `CSV.read(file, DataFrame)` is called, since columns are built only for the purpose of use in a `DataFrame`. Another related example is [`Arrow.Table`](https://arrow.juliadata.org/dev/manual/#Arrow.Table), where arrow data is inherently immutable so columns can't be accidentally mutated anyway. To be able to mutate arrow data, mutable columns must be materialized, which can be accomplished via `DataFrame(arrow_table; copycols=true)`.
```

Comment:
I would add an explanation that `copycols=nothing` copies by default unless signaled otherwise (and I added line breaks)
```suggestion
The generic single-argument constructor `DataFrame(table)` has `copycols=nothing`
by default, where column copying is performed unless Tables.jl-compatible input
signals that a copy of columns doesn't need to be made (this is done by wrapping
the source table in `Tables.CopiedColumns`).
[CSV.jl](https://csv.juliadata.org/stable) does this when `CSV.read(file, DataFrame)`
is called, since columns are built only for the purpose of use in a `DataFrame`.
Another related example is [`Arrow.Table`](https://arrow.juliadata.org/dev/manual/#Arrow.Table),
where arrow data is inherently immutable so columns can't be accidentally mutated
anyway. To be able to mutate arrow data, mutable columns must be materialized,
which can be accomplished via `DataFrame(arrow_table; copycols=true)`.
```

Reply:
```suggestion
The generic single-argument constructor `DataFrame(table)` has `copycols=nothing`
by default, meaning that columns are copied unless `table`
signals that a copy of columns doesn't need to be made (this is done by wrapping
the source table in `Tables.CopiedColumns`).
[CSV.jl](https://csv.juliadata.org/stable) does this when `CSV.read(file, DataFrame)`
is called, since columns are built only for the purpose of use in a `DataFrame`.
Another example is [`Arrow.Table`](https://arrow.juliadata.org/dev/manual/#Arrow.Table),
where arrow data is inherently immutable so columns can't be accidentally mutated
anyway. To be able to mutate arrow data, columns must be materialized,
which can be accomplished via `DataFrame(arrow_table, copycols=true)`.
```

Mentioning `CSV.read` is a bit weird since the constructor form would be `DataFrame(CSV.File(...))`. Does that make a copy or not?

Reply:
`DataFrame(CSV.File(...))` is exactly the same as `CSV.read`.

Reply:
Ah I thought `CSV.read` had been added because it was a bit more efficient. Let's show both then?
```suggestion
The generic single-argument constructor `DataFrame(table)` has `copycols=nothing`
by default, meaning that columns are copied unless `table`
signals that a copy of columns doesn't need to be made (this is done by wrapping
the source table in `Tables.CopiedColumns`).
[CSV.jl](https://csv.juliadata.org/stable) does this when `DataFrame(CSV.File(file))`
`CSV.read(file, DataFrame)` are called, since columns are built only for the purpose
of use in a `DataFrame`.
Another example is [`Arrow.Table`](https://arrow.juliadata.org/dev/manual/#Arrow.Table),
where arrow data is inherently immutable so columns can't be accidentally mutated
anyway. To be able to mutate arrow data, columns must be materialized,
which can be accomplished via `DataFrame(arrow_table, copycols=true)`.
```

Reply:
(I have updated your last proposal following the fact that `CSV.read` is the only one that is non-copying)

Reply:
@quinnj Could you explain how this works? AFAIK `DataFrame(CSV.File(...))` doesn't copy everything, it's just that `CSV.read` avoids even more copies?

Reply:
You need `Tables.CopiedColumns` wrapper, see the definition of `CSV.read`:
```
function read(source, sink=nothing; copycols::Bool=false, kwargs...)
    if sink === nothing
        throw(ArgumentError("provide a valid sink argument, like `using DataFrames; CSV.read(source, DataFrame)`"))
    end
    Tables.CopiedColumns(CSV.File(source; kwargs...)) |> sink
end
```
x-ref: https://github.com/JuliaData/CSV.jl/blob/52663d4f1edfce120ecc5c43088b52d945b4a293/src/CSV.jl#L53

Reply:
I guess the rationale is that the result of `CSV.File` could potentially be passed to different sinks if it was stored. While `CSV.read` ensures that the result of `CSV.File` is used exactly once.

Reply:
OK, let's go with this version then.